src/Routing/Duplex.purs

```diff
@@ -79,12 +89,41 @@ suffix (RouteDuplex enc dec) s = RouteDuplex (\a -> enc a <> Printer.put s) (dec
 path :: forall a b. String -> RouteDuplex a b -> RouteDuplex a b
 path = flip (foldr prefix) <<< String.split (Pattern "/")
 
+-- | Modifies given `codec` by requiring it to be prefixed with '/'.
+-- | You can think of it as stripping the '/' at the beginning of route, failing if it's not there.
+-- |
+-- |```purescript
+-- | parse (root segment) "/abc" == Right "abc"
+-- | parse (root segment) "abc" == Left (Expected "" "abc")
+-- | print (root segment) "abc" == "/abc"
+-- |```
 root :: forall a b. RouteDuplex a b -> RouteDuplex a b
 root = path ""
 
+-- | Modifies given `codec` by requiring it to consume the entire rest of input.
+-- | That is `end codec` succeeds if `codec` suceeds AND all the
+-- | remaining segments have been consumed in the process.
+-- |
+-- |```purescript
+-- | parse (end segment) "abc" == Right "abc"
+-- |
+-- | parse (end segment) "abc/def" == Left (ExpectedEndOfPath "def")
+-- |```
 end :: forall a b. RouteDuplex a b -> RouteDuplex a b
 end (RouteDuplex enc dec) = RouteDuplex enc (dec <* Parser.end)
 
+-- | Process path segment as a String.
+-- | Note that uri encoding / decoding is done automatically.
+-- |
+-- | ```purescript
+-- | parse segment "abc"       == Right "abc"
+-- | parse segment "abc%20def" == Right "abc def" -- automatic decoding of uri components
+-- | parse segment "abc/def"   == Right "abc"
+-- | parse segment "/abc"      == Right "" -- the empty string before the first '/'
```

Comment:
I like the examples! Maybe we could add an one here that uses `root`.

Reply:
From library user's perspective I always find examples useful. I'll add one with root as you suggest.
But in general I feel slightly uncomfortable without verifying all of them with some kind of "doctest" tool. Does purescript have something along those lines? I'll definitely check all the examples in the repl, but for long term maintenance having an automated way to check the examples would be better. I found https://github.com/hdgarrood/purs-doctest/ but that seems to be abandoned..